man/moonunit-lt.1
.TH MOONUNIT-LT "1" "February 2008"
.SH NAME
moonunit-lt \- \fBLibtool\fP support for \fBMoonUnit\fP
.SH SYNOPSIS
moonunit-lt [\fB--plugin\fP \fImodule\fP] [\fB--wrap\fP \fIcommand_prefix\fP]
<moonunit parameters | \fIlibtool_archive\fP>...
.SH DESCRIPTION
\fBmoonunit-lt\fP allows uninstalled libtool archive (.la) files
to be used with \fBmoonunit\fP. Libraries will be relinked as
necessary and parameters rewritten before invoking \fBmoonunit\fP.

This script allows \fBMoonUnit\fP tests to be compiled and run
as part of a project built with \fBlibtool\fP.  All types of
libraries, including static, are supported.

.SH OPTIONS
\fBmoonunit-lt\fR expects a list of parameters as appropriate for
\fBmoonunit\fR, with the exception that libtool archive files may
safely be specified.
.PP
In addition, the following options are available:
.PP
.IP \fB--plugin\fP\ \fImodule\fP
Allows \fImodule\fP (which should be a libtool archive) to be
found and loaded as a \fBMoonUnit\fP plugin.  This is useful
for testing a new plugin within its own build tree.
.IP \fB--wrap\fP\ \fIcommand_prefix\fP
Prefixes the invocation of \fBmoonunit\fP with \fIcommand_prefix\fP,
e.g. \fBgdb --args\fP.  This is used primarily for testing of
\fBmoonunit\fP itself.
.IP \fB-?\fP
Lists brief usage and option information in addition to that
provided by \fBmoonunit\fP.
.IP \fB--help\fP
Equivalent to \fB-?\fP
.SH EXAMPLES
.TP
moonunit-lt mylib.la

Run all unit tests in the libtool library mylib.la.
.TP
moonunit-lt --plugin mylogger.la -l mylogger mylib.la

Run all unit tests in the libtool library mylib.la, logging
results using the logger plugin in the libtool library
mylogger.la.
.TP
moonunit-lt --wrap valgrind mylib.la

Run all unit tests in the libtool library mylib.la, with
\fBmoonunit\fP itself being run in the memory debugger \fBvalgrind\fP.

.SH COPYRIGHT
Copyright (c) 2007-2008, Brian Koropoff.  All rights reserved.

Redistribution in source and binary form is permitted under
the terms of the 3-clause BSD license.  See COPYING in the
documentation or source distribution for details.
.SH SEE\ ALSO
\fBmoonunit\fP(1), \fBmoonunit-stub\fP(1), \fBmoonunit-strip\fP(1)


src/moonunit/moonunit-lt.sh
#!/bin/bash
#
# Copyright (c) 2007-2008 Brian Koropoff
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the name of the Moonunit project nor the
#       names of its contributors may be used to endorse or promote products
#       derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY BRIAN KOROPOFF ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL BRIAN KOROPOFF BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

dir=`dirname $0`
if [ -z "$dir" ]
then
    moonunit="moonunit"
else
    moonunit="$dir/moonunit"
fi

tempdir=`mktemp -d /tmp/moonunit-lt.XXXXXXXXXX` || exit 1
libtool --config > "$tempdir/ltconfig"
objdir=`(source "$tempdir/ltconfig" && echo $objdir)`
module_ext="so"

arg=$1
while [ -n "$arg" ]
do
    shift
    case "$arg" in
        --help|'-?')
            name=`basename $0`
            cat << __EOF__
$name -- libtool support wrapper for MoonUnit

  This script allows uninstalled libtool archive (.la) files to be run directly
  in MoonUnit (e.g. as part of 'make check' in a build tree).  Any referenced
  .la files will be relinked/redirected as necessary before passing control to
  moonunit.  Any sort of library -- static, shared, or loadable module -- should
  work.

Options ($name-specific):
  --plugin module.la          Allow loading of module.la as a MoonUnit plugin
  --wrap <prefix>             Prefixes the invocation of moonunit with <prefix>
                              (e.g. "gdb --args")

__EOF__
            command=("${command[@]}" "$arg")
            ;;
        --wrap)
            wrap="$1"
            shift
            ;;
        --plugin)
            dlopens=("${dlopens[@]}" -dlopen "$1")
            shift
            ;;
        *.la)
            # Check if a compiled dynamic module already exists
            module_file="$(dirname "$arg")/$objdir/$(basename "$arg" | sed "s/\.la$/.${module_ext}/")" 
            if [ -x "${module_file}" ]
            then
                # Just use the module directly, and add a -dlopen directive so libtool will
                # set the library path correctly
                dlopens=("${dlopens[@]}" -dlopen "$arg")
                command=("${command[@]}" "${module_file}")
            else
                # We have to relink the library into a module
                relink_name=`basename "$arg"`
                so_name=`echo "$relink_name" | sed 's/\.la$/\.so/'`
                # Relink into a module
                libtool --mode=link ${CC:-cc} -shared -module -o "$tempdir/$relink_name" -rpath "$tempdir" "$arg" >/dev/null 2>&1
                # Add it to the command
                dlopens=("${dlopens[@]}" -dlopen "$tempdir/$relink_name")
                command=("${command[@]}" "$tempdir/$objdir/$so_name")
            fi
            ;;
         *)
            command=("${command[@]}" "$arg")
            ;;
    esac
    arg=$1
done   

libtool --mode=execute "${dlopens[@]}" $wrap $moonunit "${command[@]}"
rm -rf ${tempdir}
